### ISSUE TITLE
Integration for CKAN API in SilverStripe

### ISSUE THREAD

**Issue by rupachup** [silverstripe/silverstripe-ckan-registry#1]
### User story
As a SilverStripe developer
I want to be able to make calls to CKAN instances via API in PHP
So that I can extract data and metadata for my own use inside SilverStripe framework.

### Acceptance criteria
There is a wrapper for CKAN API that uses PHP
I can write a function in SilverStripe framework syntax and get data from a CKAN instance into a format workable inside SilveStripe framework.
We attempt to evaluate and reuse an existing php wrapper if it is appropriate or we just use guzzle.

### Notes

- his will be needed for all the CMS side config of the register interface. We'll need to query a dataset and resources and get columns coulmns and metadata to allow for CMS side configuration by a CMS user.
- Useful if users want to bring in data or pieces of data to manipulate inside SilverStripe with arrays/ORM.

- PHP class wrapper for CKAN - already exist or exist for other frameworks, suggest a review and selection if they are reasonable.

- https://github.com/opencolorado/PHP-Wrapper-for-CKAN-API
- https://github.com/GSA/ckan-php-client
- https://github.com/SilexConsulting/CKAN_PHP
- https://github.com/Germanaz0/laravel-ckan-api
- https://github.com/govCMS/ckan_connect
- Allows for building wider integrations after the initial frontend register UI.

This will be needed for most of the CMS user level stories to query and return information for the CMS user to config.

Was suggested we can go loosely coupled and just use guzzle to call API and collect the metadata about the dataset.

**Comment by robbieaverill** [comment]
I read a nice article a while back about wrapping up external API responses into SilverStripe ORM style classes so they could be accessed easily via something like `CKANRegistry::get()->filter('SomeColumn', 'my value')->limit(5);` etc: https://www.bigfork.co.uk/blog/working-with-external-data-sources-in-silverstripe-part-2-querying-an-api/

**Comment by robbieaverill** [comment]
## Quick review of example repos

### https://github.com/opencolorado/PHP-Wrapper-for-CKAN-API

Standalone PHP library using a bespoke cURL wrapper to access the CKAN data. Doesn't have composer config (won't be in packagist), would probably require us to fork and maintain it.

### https://github.com/GSA/ckan-php-client

Last commit a year ago, has composer/packagist/CI, looks better structurally than the previous one. Also uses cURL. It looks like there are some debugging lines in the code which would cause syntax errors... maybe the 0.0.1 tag (only release) doesn't have those in it though.

### https://github.com/SilexConsulting/CKAN_PHP

Is available on Packagist, but looks like an inactive and rarely used repo. To be honest, this is just a Guzzle wrapper that returns the result, nothing more. I don't think we can use this.

### https://github.com/govCMS/ckan_connect

Written for Drupal/govCMS, has only has its initial commit and looked unmaintained. Is not available through Packagist. Uses Guzzle. It has a client to connect and a parser to return the package and resource ID as well as the resource view ID and URL. If we used this we'd need to fork it and remove the Drupal code.

### https://github.com/Germanaz0/laravel-ckan-api

Build for Laravel 5. Is on Packagist and uses Guzzle for the connection. It defines "repository" classes which represent parts of the API result. These classes extend a BaseRepository class which contains the Guzzle client implementation logic. There is a little bit of Laravel coupling in it, but not too much.

## Other options I found

Re: https://github.com/ckan/ckan/wiki/CKAN-API-Clients#php

### https://github.com/PeeHaa/CkanConsumer

Very lightweight, exists on Packagist and uses PHP stream functions to get the content. Doesn't have much of a following though.

### https://github.com/jeffreybarke/Ckan_client-PHP

Hasn't been contributed to for 8 years, doesn't have Packagist or composer config. Uses cURL.

---

To be honest, none of these libraries look like a great choice for us. There are two that are listed on the ckan/ckan repository's wiki, but both look old and unmaintained. The Laravel package looks like the best implementation at a quick glance. We could also fork one of these and update it a little or remove some coupling.

**Comment by ScopeyNZ** [comment]
Ok I agree that this Laravel package is a good option. To be honest I'd just want to take half the files though. There's also no license on the code. I'll try to communicate with the author of that repository now.

**Comment by ScopeyNZ** [comment]
Here's some thoughts:

- We can't forget that there's two contexts where we will be querying the API, PHP __and__ Javascript. 
- Frontend code will completely be Javascript, we don't really want SilverStripe to be a weird sort of proxy for data.govt here.
- Considering some JS library is a requirement already we should consider using that in the backend where possible
- Here's where I think we'll be querying CKAN in the backend
 - All resources for a dataset to populate the dropdown here: https://projects.invisionapp.com/share/8GOKGM7BTR5#/screens/328910661
 - All the columns and meta data for those columns: https://projects.invisionapp.com/share/8GOKGM7BTR5#/screens/325469913
 - Populating options for a dropdown filter: https://projects.invisionapp.com/share/8GOKGM7BTR5#/screens/328154359
- First and third of those can definitely use a JS API.
- The second one could use an API? We're making a gridfield...
 - Selecting a resource in the selection field could fetch all the column metadata in a blob and when saving the page it can just post it through?

TLDR;

Maybe we don't even need a PHP API for CKAN?

**Comment by camfindlay** [comment]
Something to test will be any issues with CORS headers from the CKAN API response if we do the calls via JS

A light PHP approach might be able to avoid any tricky stuff like this.
PHP wrapper might also have an advantage later if we do any buildtasks etc for future features

**Comment by ScopeyNZ** [comment]
I've been playing around with this in more detail now that I've picked up #1 and #17. No matter what we're going to have _some_ JS API. We don't want people using these pages on the front-end to be using the API with a SilverStripe site as a sort of proxy. I think there's probably still a case for both PHP & JS implementations though. We can get by with a pretty simple implementation. 

Additionally I've not bothered with the JS CKAN library listed. I've instead written a super lightweight wrapper around `isomorphic-fetch` which we use in `silverstripe-admin`. We can discuss this further in the upcoming PR. Switching to the library is definitely possible, but maybe not worth any overhead.
